### ISSUE TITLE
SMT encoding bug for equality in refinement

### ISSUE THREAD

**Issue by jkzinzindohoue** [FStarLang/FStar#771]
In the following interface file (.fsti) I have the following issue:

```
module Bla

val len: pos
val clen: l:FStar.UInt32.t{FStar.UInt32.v l = len}

val template: idx:nat{idx < len} -> Tot pos
val ctemplate: idx:nat{idx < len} -> Tot (x:FStar.UInt32.t{FStar.UInt32.v x = template idx}) // fails
```
while permuting the order of the equality in the refinement makes it go through:
```
module Bla

val len: pos
val clen: l:FStar.UInt32.t{FStar.UInt32.v l = len}

val template: idx:nat{idx < len} -> Tot pos
val ctemplate: idx:nat{idx < len} -> Tot (x:FStar.UInt32.t{template idx = FStar.UInt32.v x}) // works
```

I have trouble isolating the issue more precisely, but opening this module in another file causes similar issues (one direction of the equality works but not the other).

**Comment by catalin-hritcu** [comment]
This is causing some worrisome Z3 warnings:
```
[hritcu@detained bug-reports]$ fstar.exe bug771.fsti
unknown(0,0-0,0): (Warning) Bug771: Unexpected output from Z3: (error "line 79378 column 9: unknown function/constant Tm_refine_69847f191a94c924b2fc33a0c8d0133d")

unknown(0,0-0,0): (Warning) Bug771: Unexpected output from Z3: (error "line 79454 column 9: unknown function/constant Tm_refine_69847f191a94c924b2fc33a0c8d0133d")

unknown(0,0-0,0): (Warning) Bug771: Unexpected output from Z3: (error "line 79530 column 9: unknown function/constant Tm_refine_69847f191a94c924b2fc33a0c8d0133d")

unknown(0,0-0,0): (Warning) Bug771: Unexpected output from Z3: (error "line 79606 column 9: unknown function/constant Tm_refine_69847f191a94c924b2fc33a0c8d0133d")

unknown(0,0-0,0): (Warning) Bug771: Unexpected output from Z3: (error "line 79682 column 9: unknown function/constant Tm_refine_69847f191a94c924b2fc33a0c8d0133d")

./bug771.fsti(7,59-7,90): (Error) Unknown assertion failed
Verified i'face: Bug771 (2217 milliseconds)
1 error was reported (see above)
```
These warnings don't appear in the 2nd version of the code:
```
[hritcu@detained bug-reports]$ fstar.exe bug771b.fsti
Verified i'face: Bug771b (515 milliseconds)
All verification conditions discharged successfully
```
So it seems like a bug in the SMT encoding.

**Comment by catalin-hritcu** [comment]
The smt2 code triggering this problem looks like this:
```
;; def=/home/hritcu/Projects/fstar/pub/bin/../ulib/prims.fst(218,36-218,43); use=/home/hritcu/Projects/fstar/pub/bin/../ulib/prims.fst(218,36-218,43)
(Valid 
;; def=/home/hritcu/Projects/fstar/pub/bin/../ulib/prims.fst(218,36-218,43); use=/home/hritcu/Projects/fstar/pub/bin/../ulib/prims.fst(218,36-218,43)
(Prims.hasEq (Tm_refine_69847f191a94c924b2fc33a0c8d0133d FStar.UInt32.n
Prims.int))
)
)
```
and is the same in both queries (up to alpha renaming, `Tm_refine_69847f191a94c924b2fc33a0c8d0133d` = `Tm_refine_9443c89646beb41e40721865c562f824`). However, the second query starts with a much larger prelude, while the first only has the first 4 lines of it (up to the missing `declare_fun`):
```
; Starting query at ./bug771.fsti(9,59-9,90)
(declare-fun label_201 () Bool)
(declare-fun label_200 () Bool)
(declare-fun label_199 () Bool)
(declare-fun Tm_refine_9443c89646beb41e40721865c562f824 (Term Term) Term)
;;;;;;;;;;;;;;;;refinement kinding
(assert (!
(forall ((@x0 Term) (@x1 Term))
 (! (HasType (Tm_refine_9443c89646beb41e40721865c562f824 @x0
@x1)
Tm_type)
  
:pattern ((HasType (Tm_refine_9443c89646beb41e40721865c562f824 @x0
@x1)
Tm_type))))
:named refinement_kinding_Tm_refine_9443c89646beb41e40721865c562f824))
;;;;;;;;;;;;;;;;(i#831:int{(l_or (b2t (op_GreaterThan i@0 0)) (size i@0 n))})
(assert (!
(forall ((@u0 Fuel) (@x1 Term) (@x2 Term) (@x3 Term))
 (! (iff (HasTypeFuel @u0
@x1
(Tm_refine_9443c89646beb41e40721865c562f824 @x2
@x3))
(and (HasTypeFuel @u0
@x1
@x3)

;; def=/home/hritcu/Projects/fstar/pub/ulib/FStar.UInt.fst(24,28-24,36); use=./bug771.fsti(9,59-9,90)
(or 
;; def=/home/hritcu/Projects/fstar/pub/bin/../ulib/prims.fst(635,17-635,22); use=./bug771.fsti(9,59-9,90)
(> (BoxInt_proj_0 @x1)
(BoxInt_proj_0 (BoxInt 0)))


;; def=/home/hritcu/Projects/fstar/pub/ulib/FStar.UInt.fst(24,28-24,36); use=./bug771.fsti(9,59-9,90)
(Valid 
;; def=/home/hritcu/Projects/fstar/pub/ulib/FStar.UInt.fst(24,28-24,36); use=./bug771.fsti(9,59-9,90)
(FStar.UInt.size @x1
@x2)
)
)
))
  
:pattern ((HasTypeFuel @u0
@x1
(Tm_refine_9443c89646beb41e40721865c562f824 @x2
@x3)))))
:named refinement_interpretation_Tm_refine_9443c89646beb41e40721865c562f824))
;;;;;;;;;;;;;;;;haseq for Tm_refine_9443c89646beb41e40721865c562f824
(assert (!
(forall ((@x0 Term) (@x1 Term))
 (! (iff (Valid (Prims.hasEq (Tm_refine_9443c89646beb41e40721865c562f824 @x0
@x1)))
(Valid (Prims.hasEq @x1)))
  
:pattern ((Valid (Prims.hasEq (Tm_refine_9443c89646beb41e40721865c562f824 @x0
@x1))))))
:named haseqTm_refine_9443c89646beb41e40721865c562f824))
```

**Comment by catalin-hritcu** [comment]
The query formula leading to the bad encoding looks as follows at the F* level:
```
Encoding query formula: (l_Forall (fun idx -> (l_Forall (fun x -> (l_and (l_Forall (fun a -> (l_imp (hasEq a@0) (hasEq a@0)))) (l_and (hasEq (x#843:int{(l_or (size x@0 n) (b2t (op_GreaterThan x@0 0)))})) (l_Forall (fun x -> (b2t (op_GreaterThanOrEqual n 0))))))))))
``` 
The corresponding query formula in the 2nd file is this:
```
Encoding query formula: (l_Forall (fun idx -> (l_Forall (fun x -> (l_and (l_Forall (fun a -> (l_imp (hasEq a@0) (hasEq a@0)))) (l_and (hasEq (i#831:int{(l_or (b2t (op_GreaterThan i@0 0)) (size i@0 n))})) (l_Forall (fun i -> (l_imp (l_not (b2t (op_GreaterThan i@0 0))) (b2t (op_GreaterThanOrEqual n 0)))))))))))
```
Don't see and difference that would justify the later problems.

**Comment by catalin-hritcu** [comment]
@nikswamy Could this be caused by a wrong cache state in the following code:
https://github.com/FStarLang/FStar/blob/master/src/smtencoding/encode.fs#L571

**Comment by aseemr** [comment]
The root cause of this bug are some missing definitions in the smt file. Running regressions on the fix.

**Comment by aseemr** [comment]
I also have a hypothesis on why switching the arguments in the refinement works, but need to validate the hypothesis first.

**Comment by aseemr** [comment]
Fixed with a8b98ae66358576b994107e698999bbec1ea9d55. There are no errors in the encoding and verification also goes through.
